source/tutorials/integration-testing-using-virtual-machines.rst

```diff
@@ -0,0 +1,228 @@
+          serviceConfig.User = username;
+        };
+      };
+
+      client = { 
+        imports = [ sharedModule ];
+      };
+    };
+
+    testScript = ''
+      import json
+      import sys
+
+      start_all()
+
+      server.wait_for_open_port(${toString postgrestPort})
+
+      expected = [
+          {"id": 1, "done": False, "task": "finish tutorial 0", "due": None},
+          {"id": 2, "done": False, "task": "pat self on back", "due": None},
+      ]
+
+      actual = json.loads(
+          client.succeed(
+              "${pkgs.curl}/bin/curl http://server:${toString postgrestPort}/${table}"
+          )
+      )
+
+      if expected != actual:
+          sys.exit(1)
+    '';
+  })
+
+A few notes:
+
+- Between the machines defined inside the ``nodes`` attribute, hostnames 
+  are resolved based on their attribute names. In this case we have ``client`` and ``server``.
+
+- The testing framework exposes a wide set of operations used inside the ``testScript``.
+  A full set of testing operations is part of  
+  `VM testing operations API Reference <https://nixos.org/manual/nixos/stable/index.html#sec-nixos-tests>`_.
+
+
+Running tests
+-------------
+
+To set up all machines and execute the test script:
+
+.. code:: shell-session 
+
+  $ nix-build postgrest.nix
+
+You'll notice an error message if something goes wrong.
+
+In case the tests succeed, you should see at the end:
+
+.. code:: shell-session
+
+  1 out of 1 tests succeeded
+
+
+Developing and debugging tests
+------------------------------
+
+When developing tests or when something breaks, it's useful to interactively fiddle
+with the script or access a terminal for a machine.
+
+To interactively start a Python session with testing framework:
+
+.. code:: shell-session 
+
+  $ $(nix-build -A driver postgrest.nix)/bin/nixos-test-driver
+  ...
+  starting VDE switch for network 1
+  >>>
+
+You can run `any of the testing operations <https://nixos.org/manual/nixos/stable/index.html#sec-nixos-tests>`_. 
+``testScript`` attribute from our ``postgrest.nix`` definition can be executed with ``test_script()`` function.
+
+To start all machines and enter a telnet terminal to a specific machine:
+
+.. code:: shell-session 
+
+   >>> start_all()
+   ...
+   >>> server.shell_interact()
+   server: Terminal is ready (there is no prompt):
+
+   uname -a 
+   Linux server 5.10.37 #1-NixOS SMP Fri May 14 07:50:46 UTC 2021 x86_64 GNU/Linux
+
+
+Next steps
+----------
+
+- Running integration tests on CI requires hardware acceleration, which many CIs do not support. 
+  To run integration tests on :ref:`GitHub Actions <github-actions>` see 
+  `how to disable hardware acceleration <https://github.com/cachix/install-nix-action#how-can-i-run-nixos-tests>`_.
+
+- NixOS comes with a large set of tests that serve also as educational examples, a good inspiration is `Matrix bridging with an IRC <https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/matrix-appservice-irc.nix>`_.
```

Comment:
my personal favorite example is the bittorrent test. i can also provide a nice picture that shows what kind of complexity the test maintains with few lines that i would share if you are interested.

![Screenshot from 2021-06-03 14-18-08](https://user-images.githubusercontent.com/29044/120643600-a9452c80-c476-11eb-9379-47c798e9c3b8.png)

Reply:
That one is pretty huge, I'd like to give smaller examples that demonstrate the power of what it can do.